### PR TITLE
Batching Strategy for SimpleGarbageCollector Fix #1543

### LINKED ISSUE #1543: Use batching strategy for SimpleGarbageCollector candidate memory utilization
The SimpleGarbageCollector tries to gather, check, and delete candidate files for garbage collection without running out of memory. It currently tries to do this by monitoring its own memory utilization when filling up its internal data structure of candidates, and then processing them before memory is full (or near full), clearing, and resuming collection of more candidates.

This makes the SGC handle as many candidates as possible in each pass. However, it makes memory utilization less predictable, and unstable.

A similar problem occurred in trying to test Upgrader9to10's memory utilization and was fixed in #1441 by adopting a strategy of batching candidates in approximately 8MB chunks rather than runtime memory monitoring.

The SimpleGarbageCollector should be changed to employ a similar batching strategy. The batch size could be configurable to give system administrators more control over maximum memory utilization by the SGC.

### PR DESCRIPTION
**Pull request by Manno15**
Solution for #1543. Takes @ctubbsii  batching changes on #1441 and applying it to the SimpleGarbageCollector. I am not fully confident on my test for this feature so feadback on that will be much appreciated.

### CODE REVIEW

File: server/gc/src/main/java/org/apache/accumulo/gc/SimpleGarbageCollector.java [hunk 1]
```diff
@@ -669,14 +665,13 @@ private HostAndPort startStatsService() {
...
+   * @param candidateLength
+   *          Candidate Length
    * @return true if system is almost out of memory
-   * @see #CANDIDATE_MEMORY_PERCENTAGE
+   * @see #CANDIDATE_BATCH_SIZE
    */
-  static boolean almostOutOfMemory(Runtime runtime) {
-    return runtime.totalMemory() - runtime.freeMemory()
-        > CANDIDATE_MEMORY_PERCENTAGE * runtime.maxMemory();
+  static boolean almostOutOfMemory(Long candidateLength) {
```
Comment (ivakegg) [inline server/gc/src/main/java/org/apache/accumulo/gc/SimpleGarbageCollector.java:673]: I don't think the method comment is true anymore.  Perhaps this method name and comments should be changed altogether.

Comment (Manno15) [reply server/gc/src/main/java/org/apache/accumulo/gc/SimpleGarbageCollector.java:673]: That's fair, I wasn't quite sure on that either. Do you have any suggestions?

Comment (keith-turner) [reply server/gc/src/main/java/org/apache/accumulo/gc/SimpleGarbageCollector.java:673]: The method is simple now, I would inline it and remove the associated test.

File: server/gc/src/main/java/org/apache/accumulo/gc/SimpleGarbageCollector.java [hunk 2]
```diff
@@ -209,20 +205,20 @@ public boolean getCandidates(String continuePoint, List<String> result)
...
-
-        result.add(cand);
-        if (almostOutOfMemory(Runtime.getRuntime())) {
-          log.info("List of delete candidates has exceeded the memory"
+        String candidate = candidates.next();
+        candidateLength += candidate.length();
+        result.add(candidate);
+        if (almostOutOfMemory(candidateLength)) {
+          log.info("List of delete candidates has exceeded the batch size"
               + " threshold. Attempting to delete what has been gathered so far.");
```
Comment (keith-turner) [inline server/gc/src/main/java/org/apache/accumulo/gc/SimpleGarbageCollector.java:218]: May be useful to include the batch size.  This probably needs to be reformatted.

```suggestion
          log.info("Candidate batch of size {} has exceeded the"
              + " threshold. Attempting to delete what has been gathered so far.", candidateLength);
```

File: server/gc/src/main/java/org/apache/accumulo/gc/SimpleGarbageCollector.java (hunk 1 above)
Comment (Manno15) [reply server/gc/src/main/java/org/apache/accumulo/gc/SimpleGarbageCollector.java:673]: That is what I originally had but was unsure on if removing the test was a good idea.

**Comment by Manno15** [comment]
@keith-turner Made your requested changes.

**Comment by ctubbsii** [comment]
@keith-turner Please don't forget to set the project version before the issue or PR is closed, so it shows up on the appropriate project planning board.

**Comment by ctubbsii** [comment]
@Manno15 BTW, thanks for this! I didn't have a chance to look at the substance of this change before now. This is a nice improvement, in terms of having consistent memory utilization in accumulo-gc and ensuring it always makes progress.

**Pull request merged**